pydantic/funcs.py

```diff
@@ -0,0 +1,301 @@
+from __future__ import annotations
+
+import sys
+from typing import Any, Generic, Iterable, TypeVar, overload
+
+from pydantic_core import CoreConfig, SchemaSerializer, SchemaValidator, core_schema
+
+from pydantic.json_schema import DEFAULT_REF_TEMPLATE, GenerateJsonSchema
+
+from ._internal import (
+    _generate_schema,
+    _typing_extra,
+)
+
+T = TypeVar('T')
+
+
+class TypeMetadata(Generic[T]):
+    def __init__(self, core_schema: core_schema.CoreSchema, core_config: CoreConfig) -> None:
+        self.core_schema = core_schema
+        self.core_config = core_config
+
+
+@overload
+def analyze_type(__type: type[T], *, config: CoreConfig | None = ...) -> TypeMetadata[T]:
+    ...
+
+
+@overload
+def analyze_type(__type: Any, *, config: CoreConfig | None = ...) -> TypeMetadata[Any]:
+    ...
+
+
+def analyze_type(
```

Comment:
I'm open to other naming or changes here, but I think in general it makes sense to have a single function to do the type analysis and be able to feed that into the other functions.